File: evals/metrics.ts
/**
 * Métricas de retrieval puras, sin side effects.
 *
 * Documentación corta para que cualquiera pueda auditar la fórmula:
 * - recall@k: cuántos chunks relevantes aparecen en el top-k recuperado,
 *   normalizado por el total de chunks relevantes conocidos.
 * - MRR: posición inversa del primer chunk relevante. Penaliza tener la
 *   respuesta abajo en el ranking aunque esté presente.
 * - hit@k: 1 si hay al menos un chunk relevante en el top-k, 0 si no.
 */

export function recallAtK(
  retrievedIds: readonly string[],
  groundTruthIds: readonly string[],
  k: number
): number {
  if (groundTruthIds.length === 0) {
    return 0;
  }

  const topK = new Set(retrievedIds.slice(0, k));
  let hits = 0;
  for (const id of groundTruthIds) {
    if (topK.has(id)) hits += 1;
  }
  return hits / groundTruthIds.length;
}

export function meanReciprocalRank(
  retrievedIds: readonly string[],
  groundTruthIds: readonly string[]
): number {
  if (groundTruthIds.length === 0) {
    return 0;
  }

  const groundTruth = new Set(groundTruthIds);
  for (let i = 0; i < retrievedIds.length; i += 1) {
    if (groundTruth.has(retrievedIds[i])) {
      return 1 / (i + 1);
    }
  }
  return 0;
}

export function hitRateAtK(
  retrievedIds: readonly string[],
  groundTruthIds: readonly string[],
  k: number
): number {
  if (groundTruthIds.length === 0) {
    return 0;
  }

  const groundTruth = new Set(groundTruthIds);
  for (let i = 0; i < Math.min(k, retrievedIds.length); i += 1) {
    if (groundTruth.has(retrievedIds[i])) {
      return 1;
    }
  }
  return 0;
}

export function average(values: readonly number[]): number {
  if (values.length === 0) return 0;
  const sum = values.reduce((acc, n) => acc + n, 0);
  return sum / values.length;
}

export function percentile(values: readonly number[], p: number): number {
  if (values.length === 0) return 0;
  const sorted = [...values].sort((a, b) => a - b);
  const index = Math.min(sorted.length - 1, Math.floor((p / 100) * sorted.length));
  return sorted[index];
}

File: evals-py/evals_py/metrics.py
"""Métricas de retrieval puras, sin side effects.

Port EXACTO de ``evals/metrics.ts``. Los tests de ``tests/test_metrics.py``
replican los casos de ``evals/__tests__/metrics.test.ts`` para garantizar paridad
numérica entre los dos harnesses.
"""

from __future__ import annotations

from collections.abc import Sequence
import math


def recall_at_k(
    retrieved_ids: Sequence[str], ground_truth_ids: Sequence[str], k: int
) -> float:
    """recall@k: fracción de chunks relevantes presentes en el top-k."""
    if len(ground_truth_ids) == 0:
        return 0.0

    top_k = set(retrieved_ids[:k])
    hits = sum(1 for gid in ground_truth_ids if gid in top_k)
    return hits / len(ground_truth_ids)


def mean_reciprocal_rank(
    retrieved_ids: Sequence[str], ground_truth_ids: Sequence[str]
) -> float:
    """MRR: inverso de la posición (1-indexed) del primer relevante."""
    if len(ground_truth_ids) == 0:
        return 0.0

    ground_truth = set(ground_truth_ids)
    for i, rid in enumerate(retrieved_ids):
        if rid in ground_truth:
            return 1.0 / (i + 1)
    return 0.0


def hit_rate_at_k(
    retrieved_ids: Sequence[str], ground_truth_ids: Sequence[str], k: int
) -> float:
    """hit@k: 1 si hay al menos un relevante en el top-k, 0 si no."""
    if len(ground_truth_ids) == 0:
        return 0.0

    ground_truth = set(ground_truth_ids)
    for i in range(min(k, len(retrieved_ids))):
        if retrieved_ids[i] in ground_truth:
            return 1.0
    return 0.0


def average(values: Sequence[float]) -> float:
    if len(values) == 0:
        return 0.0
    return sum(values) / len(values)


def percentile(values: Sequence[float], p: float) -> float:
    """Mismo cálculo que ``evals/metrics.ts`` (índice truncado, no interpolado)."""
    if len(values) == 0:
        return 0.0
    ordered = sorted(values)
    index = min(len(ordered) - 1, math.floor((p / 100) * len(ordered)))
    return ordered[index]
